[A H M Thikhina Induwara Weragama Bakmeedeniya] iP

thikhinab opened the pull request:
# Luca
> "Efficiency is doing better what is already being done" - Peter Drucker [(source)](https://www.azquotes.com/quote/534901)

Power your life with **Luca** _chat bot_! 🔥
### What is it?
1. command line based chat bot
2. easy to use commands
3. ~~DULL~~ INTERACTIVE responses

### Commands
- `list` : list all the tasks
- `bye` : exits the chat bot
- `todo` : creates a ToDo task
- `deadline` : creates a Deadline task along with `/by` keyword 
- `event` : creates a Event task along with `/at` keyword
- `mark` : marks the task as done
- `unmark` : marks he task as not done 
- `find` : list out tasks with keyword

### Look out for
- [x] Date Compatibility
- [x] Save Task locally
- [ ] GUI

### Explore the code
We welcome Code Enthusiasts to take a look into inner workers of Luca!
Here is a peek into the logic of `execute` method in `MarkCommad` class. 
```java
    public void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException {
        Task task = taskList.get(point - 1); // point: 1-based index of task in the list
        task.markAsDone();
        storage.saveToFile(taskList);
        ui.showMessage("Great! I have marked this task as done:\n " + task);
    }
````

Commit:
Add Gradle support

thikhinab committed:
Implemented the skeletal version of Duke

thikhinab committed:
Added the list feature

thikhinab committed:
Implemented facility to mark and unmark task as done.

thikhinab committed:
Fixed bug in adding tasks.

thikhinab committed:
Updated input test cases and expected outputs.

thikhinab committed:
Created ToDo, Deadline and Event Classes

thikhinab committed:
Created functionality for ToDos, Deadline and Event tasks.

thikhinab committed:
Updated the test cases and expected outputs.

thikhinab committed:
Created Exceptions classes.

thikhinab committed:
Handled Exceptions.

thikhinab committed:
Use of Java Collections classes

thikhinab committed:
Implemented the delete functionality.

thikhinab committed:
Introduced Enumerations for tasks.

thikhinab committed:
Debuged an deadline/event issue and updated formatting.

thikhinab committed:
Updated test cases and expected output.

thikhinab committed:
Implement the Save functionality

thikhinab committed:
Added Date and Time functionality

thikhinab committed:
Merge branch 'branch-Level-7'

thikhinab committed:
Resolve merge conflicts

thikhinab committed:
Make more OOP

thikhinab committed:
Update test input and expected output

thikhinab committed:
Organize into packages

thikhinab committed:
Change print format of event and deadline dates

thikhinab committed:
Add JUnit tests

thikhinab committed:
Update JavaDoc

thikhinab committed:
Update code to follow coding standard

thikhinab committed:
Implement find command

thikhinab committed:
Merge branch 'branch-A-JavaDoc'

thikhinab committed:
Resolve merge conflict

thikhinab committed:
Merge branch 'branch-Level-9'

thikhinab committed:
Merge to add gradle support

DavidTan0527 commented on src/main/java/duke/command/CreateDeadlineCommand.java:
```diff
@@ -0,0 +1,47 @@
+/**
+ * Responsible for the functionality needed when creating a Deadline task.
+ */
+public class CreateDeadlineCommand extends Command{
```
```suggestion
public class CreateDeadlineCommand extends Command {
```

thikhinab replied:
Thanks pointing it out!

DavidTan0527 commented on src/main/java/duke/command/CreateEventCommand.java:
```diff
@@ -0,0 +1,48 @@
+/**
+ * Responsible for the functionality needed when creating an Event task.
+ */
+public class CreateEventCommand extends Command{
```
```suggestion
public class CreateEventCommand extends Command {
```

DavidTan0527 commented on src/main/java/duke/command/CreateToDoCommand.java:
```diff
@@ -0,0 +1,44 @@
+/**
+ * Responsible for the functionality needed when creating a ToDoo task.
+ */
+public class CreateToDoCommand extends Command{
```
```suggestion
public class CreateToDoCommand extends Command {
```

DavidTan0527 commented on src/main/java/duke/command/ExitCommand.java:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Responsible for the functionality needed when exiting the chat bot.
+ */
+public class ExitCommand extends Command{
```
```suggestion
public class ExitCommand extends Command {
```

DavidTan0527 commented on src/main/java/duke/command/FindCommand.java:
```diff
@@ -0,0 +1,51 @@
+        StringBuilder stringBuilder = new StringBuilder();
+        stringBuilder.append("Here are the matching tasks in your list:\n");
+        int startIndex = 0;
+        for(int i = 0; i < taskList.size(); i++) {
```
```suggestion
        for (int i = 0; i < taskList.size(); i++) {
```

thikhinab replied:
Will correct this! Thanks!

DavidTan0527 commented on src/main/java/duke/command/ListCommand.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * Responsible for the functionality needed when listing out all the task.
+ */
+public class ListCommand extends Command{
```
```suggestion
public class ListCommand extends Command {
```

DavidTan0527 commented on src/main/java/duke/command/MarkCommand.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Responsible for the functionality needed when marking a task.
+ */
+public class MarkCommand extends Command{
```
```suggestion
public class MarkCommand extends Command {
```

DavidTan0527 commented on src/main/java/duke/command/UnmarkCommand.java:
```diff
@@ -0,0 +1,42 @@
+/**
+ * Responsible for the functionality needed when un-marking a Task.
+ */
+public class UnmarkCommand extends Command{
```
```suggestion
public class UnmarkCommand extends Command {
```

DavidTan0527 commented on src/main/java/duke/common/DukeException.java:
```diff
@@ -0,0 +1,17 @@
+ * Exceptions that are specific to the Duke ChatBot.
+ * Inherits from Exception.
+ */
+public class DukeException extends Exception{
```
```suggestion
public class DukeException extends Exception {
```

DavidTan0527 commented on src/main/java/duke/command/CommandType.java:
```diff
@@ -0,0 +1,16 @@
+/**
+ * Represents the type of command.
+ */
+public enum CommandType {
```
Great use of `enum` for handling command types!

DavidTan0527 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,299 @@
+    public static Command parse(String input) throws DukeException {
+        String[] tokens = input.split("\\s+");
+        Command command;
+            switch (tokens[0]) {
+            case "list":
+                command = new ListCommand();
+                break;
+            case "mark":
+                command = parseMarkCommand(tokens);
+                break;
+            case "unmark":
+                command = parseUnmarkCommand(tokens);
+                break;
+            case "todo":
+                command = parseTodoCommand(tokens);
+                break;
+            case "deadline":
+                command = parseDeadlineCommand(tokens);
+                break;
+            case "event":
+                command = parseEventCommand(tokens);
+                break;
+            case "delete":
+                command = parseDeleteCommand(tokens);
+                break;
+            case "bye":
+                command = new ExitCommand();
+                break;
+            case "find":
+                command = parseFindCommand(tokens);
+                break;
+            default:
+                throw new UnkownCommandException();
+            }
```
```suggestion
         switch (tokens[0]) {
         case "list":
             command = new ListCommand();
             break;
         case "mark":
             command = parseMarkCommand(tokens);
             break;
         case "unmark":
             command = parseUnmarkCommand(tokens);
             break;
         case "todo":
             command = parseTodoCommand(tokens);
             break;
         case "deadline":
             command = parseDeadlineCommand(tokens);
             break;
         case "event":
             command = parseEventCommand(tokens);
             break;
         case "delete":
             command = parseDeleteCommand(tokens);
             break;
         case "bye":
             command = new ExitCommand();
             break;
         case "find":
             command = parseFindCommand(tokens);
             break;
         default:
             throw new UnkownCommandException();
         }
```

thikhinab replied:
Thanks for pointing this out!

DavidTan0527 commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,75 @@
+    private Storage storage;
+
+    /** Temporary store of list of tasks. */
+    private TaskList tasks;
```
Great use of the plural noun `tasks` to show that `TaskList` encapsulates multiple `Task` objects.

thikhinab replied:
Thank you!

DavidTan0527 commented on src/main/java/duke/parser/Parser.java:
```diff
@@ -0,0 +1,299 @@
+                            + "start and end times in the following format:\n"
+                            + " yyyy-mm-dd HHMM HHMM in the 24 hour format");
+                }
+                i+=2;
```
```suggestion
                i += 2;
```

thikhinab replied:
Will correct the styling! Thanks!

DavidTan0527 reviewed (commented):
Overall great choice of names. Thumbs up for correct usage of cases for different types of names. Just a few spacing and indentations to fix 😄

thikhinab committed:
Correct styling

thikhinab committed:
Add CheckStyle

thikhinab committed:
Correct styling

thikhinab committed:
Implement GUI

thikhinab committed:
Merge branch 'branch-Level-10'

thikhinab committed:
Incorporated Varargs

thikhinab committed:
Improve styling

xpzmichael commented on src/main/java/luca/task/TaskList.java:
```diff
@@ -0,0 +1,107 @@
+public class TaskList {
+
+    /** List of task. */
+    private ArrayList<Task> taskList;
```
Maybe consider making taskList **final**.

xpzmichael commented on src/main/java/luca/parser/Parser.java:
```diff
@@ -0,0 +1,300 @@
+        case "delete":
+            command = parseDeleteCommand(tokens);
+            break;
+        case "bye":
+            command = new ExitCommand();
+            break;
```
If user entered **"bye do somthing /at Sunday"**, your program will end directly. Maybe it can be recognized as an "invalid user input (user typo)" or to double confirm whether user wants to quit or not.

xpzmichael commented on src/main/java/luca/task/Deadline.java:
```diff
@@ -0,0 +1,66 @@
+public class Deadline extends Task {
+
+    /** String representation of the deadline. */
+    protected LocalDateTime by;
```
The variable name "**by**" can be replaced with words like "time" as it is better to name the fields as **nouns**.

xpzmichael commented on src/main/java/luca/task/Event.java:
```diff
@@ -0,0 +1,81 @@
+     */
+    public Event(String description, LocalDateTime start, LocalDateTime end) {
+        super(description, TaskType.EVENT);
+        this.start = start;
+        this.end = end;
```
Consider changing **"start"** and **"end"** to **"startTime"** and **"endTime"**.

xpzmichael commented on src/main/java/luca/task/Event.java:
```diff
@@ -0,0 +1,81 @@
+public class Event extends Task {
+
+    /** String representation of the event start dates/times. */
+    protected LocalDateTime start;
```
Since once a task is created, time cannot be changed, the time of event/deadline can be declared final.

xpzmichael reviewed (commented):
Nice code! Very well written and well organized. Some fields can be declared final since they are not going to be changed.

thikhinab committed:
Update build specifications

thikhinab committed:
Improve list command functionality

thikhinab committed:
Add assertions

Assertions are not enforced.

Assertions could be used to find the correctness of any assumptions
made by the developer.

Insert assertions to methods where assumptions have been made.

Hence, wrong assumptions can be caught.

thikhinab committed:
Implement Streams

thikhinab committed:
Improve code quality

Quality of the code can be improved.

Improvement of code quality will improve the readability of the code.
It will also help collaborators to understand the code easily.

Let's
* use constants for recurring literals.
* use gaurd clauses.
* use the KISS principle.

thikhinab committed:
Merge pull request #1 from thikhinab/branch-A-CodeQuality

Improve code quality

thikhinab committed:
Merge pull request #2 from thikhinab/branch-A-Assertions

Add assertions

thikhinab committed:
Merge branch 'master' into branch-A-Streams

thikhinab committed:
Merge pull request #3 from thikhinab/branch-A-Streams

Implement Streams

thikhinab committed:
Set up CI

thikhinab committed:
Create gradle.yml

thikhinab committed:
Improve styling

thikhinab committed:
Implement sort functionality

thikhinab committed:
Merge branch 'branch-C-Sort'

thikhinab committed:
Improve readability

thikhinab committed:
Improve GUI

thikhinab committed:
Improve styling

thikhinab committed:
Add user guide

thikhinab committed:
Set theme jekyll-theme-minimal

thikhinab committed:
Set theme jekyll-theme-cayman

thikhinab committed:
Set theme jekyll-theme-cayman

thikhinab committed:
Set theme jekyll-theme-slate

thikhinab committed:
Fix GUI bug

thikhinab committed:
Add guard clauses

thikhinab committed:
Fix file path bug

thikhinab committed:
Merge branch 'master' of https://github.com/thikhinab/ip

thikhinab committed:
Handle edge cases

thikhinab committed:
Set theme jekyll-theme-midnight

thikhinab committed:
Change formatting of user guide

thikhinab committed:
Set theme jekyll-theme-slate

thikhinab committed:
Fix typos in user guide

thikhinab committed:
Update examples in user guide

thikhinab committed:
Update event and deadline examples in user guide
